README.md

<!-- README.md is generated from README.Rmd-->

<p align="right">

<img src="man/figures/logo.png" width="250px"></img>

</p>

<br/>

# osmenrich: enrich geocoded data using OpenStreetMap

<!-- badges: start -->

[![Github Action
test](https://github.com/sodascience/osmenrich/workflows/R-CMD-check/badge.svg)](https://github.com/sodascience/osmenrich/actions)
[![DOI](https://zenodo.org/badge/337555188.svg)](https://zenodo.org/badge/latestdoi/337555188)
<!-- badges: end -->

The goal of `osmenrich` is to easily enrich geocoded data
(`latitude`/`longitude`) with geographic features from OpenStreetMap
(OSM). The main language of the package is `R` and this package is
designed to work with the [`sf`](https://r-spatial.github.io/sf/) and
[`osmdata`](https://cran.r-project.org/web/packages/osmdata/vignettes/osmdata.html)
packages for collecting and manipulating geodata.

## Installation

To install the package, you first need to have the `remotes` package
installed. If you do not have this package yet, please install it first
with:

``` r
install.packages("remotes")
```

If you do have this package, due to recent changes in GitHub’s naming of
branches, please make sure you have the latest version of `remotes` or
at least version `2.2`.

Once you did this, to continue the installation of the `osmenrich`
package, run:

``` r
remotes::install_github("sodascience/osmenrich@main")
```

or, for the development version, run:

``` r
remotes::install_github("sodascience/osmenrich@develop")
```

This will use the default public APIs for OSM data and routing (for
computing driving/walking distances and durations). **Do not use
`osmenrich` with default APIs for large datasets\!** If you want to
learn how to use `osmenrich` for large queries follow the instructions
in section [Local API Setup](#local-api-setup) below.

## Usage

### Simple enrichment example

Let’s enrich a spatial (`sf`) dataset (`sf_example`) with the number of
waste baskets in a radius of 500 meters from each of the point specified
in a dataset:

``` r
# Import libraries
library(tidyverse)
library(sf)
library(osmenrich)

# Create an example dataset to enrich
sf_example <-
  tribble(
    ~person,  ~lat,   ~lon,
    "Alice",  52.12,  5.09,
    "Bob",    52.13,  5.08,
  ) %>%
  sf::st_as_sf(
    coords = c("lon", "lat"),
    crs = 4326
  )

# Print it
sf_example
#> Simple feature collection with 2 features and 1 field
#> geometry type:  POINT
#> dimension:      XY
#> bbox:           xmin: 5.08 ymin: 52.12 xmax: 5.09 ymax: 52.13
#> CRS:            EPSG:4326
#> # A tibble: 2 x 2
#>   person     geometry
#> * <chr>   <POINT [°]>
#> 1 Alice  (5.09 52.12)
#> 2 Bob    (5.08 52.13)
```

To enrich the `sf_example` dataset with “waste baskets” in a 500m
radius, you can create a query using the `enrich_osm()` function. This
function uses the bounding box created by the points present in the
example dataset and searches for the specified `key = "amenity"` and
`value = "waste_basket`. You can also add a custom `name` for the newly
created column and specify the radius (`r`) used in the search. See [Map
Features on the website of
OSM](https://wiki.openstreetmap.org/wiki/Map_features) for a complete
list of `key` and `value` combinations.

``` r
# Simple OSMEnrich query
sf_example_enriched <- sf_example %>%
  enrich_osm(
    name = "n_waste_baskets",
    key = "amenity",
    value = "waste_basket",
    r = 500
  )
#> Downloading data for waste_baskets... Done.
#> Downloaded 147 points, 0 lines, 0 polygons, 0 mlines, 0 mpolygons.
#> Computing distance matrix for n_waste_baskets...Done.
```

The resulting enriched dataset `sf_example_enriched` is a `sf` object
and can be printed as usual to inspect the newly added column
`n_waste_baskets`.

``` r
sf_example_enriched
#> Simple feature collection with 2 features and 2 fields
#> geometry type:  POINT
#> dimension:      XY
#> bbox:           xmin: 5.08 ymin: 52.12 xmax: 5.09 ymax: 52.13
#> geographic CRS: WGS 84
#> # A tibble: 2 x 3
#>   person     geometry n_waste_baskets
#> * <chr>   <POINT [°]>         <int>
#> 1 Alice  (5.09 52.12)            75
#> 2 Bob    (5.08 52.13)             1
```

The waste baskets column is now the result of summing all the
wastebaskets in a 500 meter radius for Alice and Bob:

![](man/figures/example_wastebaskets_r500.png)

## Local API setup

OSM enrichment can ask for a lot of data, which can overload public
APIs. If you intend to enrich large amounts of data or compute routing
distances (e.g., driving duration) between many points, you should set
up a local API endpoint.

Multiple `docker-compose` workflows for doing this are avaialble in the
separate [osmenrich\_docker
repository](https://github.com/sodascience/osmenrich_docker). Use the
`README` in the repository to select the workflow that fits your desired
outcome.

<img src="man/figures/docker.png" alt="Docker logo" width="250px"/>

<!-- CONTRIBUTING -->

## Contributing

Contributions are what make the open source community an amazing place
to learn, inspire, and create. Any contributions you make are **greatly
appreciated**.

In this project, the [Gitflow
workflow](https://nvie.com/posts/a-successful-git-branching-model/) is
used. Instead of having a single `main` branch, the project makes use of
two branches to record the history of the project: `develop` and `main`.
The `main` branch is used only for the official releases of the project,
while the `develop` branch is used to integrate the new features
developed. Finally, `feature` branches are used to develop new features
or additions to the project that will be `rebased and squashed` in the
`develop` branch.

The workflow to contribute with Gitflow becomes:

1.  Fork the Project
2.  Create your Feature Branch (`git checkout -b
    feature/<AmazingFeature>`)
3.  Commit your Changes (`git commit -m 'Add some <AmazingFeature>'`)
4.  Push to the Branch (`git push origin feature/<AmazingFeature>`)
5.  Open a Pull Request

## License and citation

The `osmenrich` package is published under the MIT license. When using
`osmenrich` for academic work, please cite:

    van Kesteren, Erik-Jan, Vida, Leonardo, de Bruin, Jonathan, & Oberski, Daniel. (2021, February 11).
    Enrich sf Data with Geographic Features from OpenStreetMaps (Version v1.0). Zenodo. http://doi.org/10.5281/zenodo.4534188

<!-- CONTACT -->

## Contact

This package is developed and maintained by the [ODISSEI Social Data
Science (SoDa)](https://odissei-data.nl/nl/soda/) team.

Do you have questions, suggestions, or remarks? File an issue in the
issue tracker or feel free to contact [Erik-Jan van
Kesteren](https://github.com/vankesteren)
(\[@ejvankesteren\](<https://twitter.com/ejvankesteren>)) or [Leonardo
Vida](https://github.com/leonardovida)
(\[@leonardojvida\](<https://twitter.com/leonardojvida>))

<img src="man/figures/word_colour-l.png" alt="SoDa logo" width="250px"/>


NEWS.md
# osmenrich

## osmenrich 1.2

### Major changes:
- Documentation updates:
  - Add hyperlink to Map Features on the OSM website
  - Remove words like 'our' and 'we'
  - Fixed documentation inaccuracies
  - Refactored naming of variables
  - Replace `dontrun` by `donttest`
  - Clarify title of `enrich_osm.R`
  - Update README and add wastebasket figure
- Package updates:
  - Added the example dataset `common_swift`
  - Move details of `enrich_opq.R` to `enrich_osm.R`
  - Add feature geometry correction in `enrich_osm.R`

## osmenrich 1.1

### Major changes:
- Documentation updates:
  - Added new tutorial _Introduction to kernels_
  - Fixed documentation links
  - Corrected documentation mistakes
- Package updates:
  - Added possibility of changing kernels' aggregation function by specifying
  the variable `reduce_fun`

## osmenrich 1.0

### Major changes:
- First release


R/enrich_osm.R
#' @name enrich
#' @title Enrich `sf` object with OSM data
#' @description Perform enriched query on OSM and add as new column.
#'
#' @param name the column name of the feature to be added
#' @param dataset target `sf` dataset to enrich with this package
#' @param key target OSM feature key to add, see [osmdata::add_osm_feature()]
#' @param value target value for OSM feature key to add, see
#'   [osmdata::add_osm_feature()]
#' @param type `character` the osm feature type or types to consider
#' (e.g., points, polygons), see details
#' @param measure `character` the measure metric used, see details
#' @param kernel `function` the kernel function used, see details
#' @param r The search radius used by the `kernel` function.
#' @param reduce_fun The aggregation function used by the `kernel` function to
#'   aggregate the retrieved data objects
#' @param control The list with configuration variables for the OSRM server.
#'   It contains `timeout`, defining the number of seconds before the request
#'   to OSRM times out, and `memsize`, defining the maximum size of the query to
#'   OSRM.
#' @param .verbose `bool` whether to print info during enrichment
#' @param ... Additional parameters to be passed into the OSM query, such as
#'   a user-defined kernel.
#'
#' @details `Type` represents the feature type to be considered. Usually this
#'   would be points, but polygons and multipolygons are also possible. This
#'   argument can also be a vector of multiple types. Non-point types will be
#'   converted to points using the `st_centroid` function from the `sf` package
#'   (NB this does not necessarily work well for all features!).
#'   Available options are:
#'   - points
#'   - lines
#'   - polygons
#'   - multilines
#'   - multipolygons
#'
#'  `Measure` represents the metric used to compute the distances or durations
#'   between the rows in the dataset and the OSM features. The following metrics
#'   are available in this package, assuming that the OSRM server is setup as
#'   suggested in our guide at:
#'   https://github.com/sodascience/osmenrich_docker:
#'   - spherical
#'   - distance_by_foot
#'   - duration_by_foot
#'   - distance_by_car
#'   - duration_by_car
#'   - distance_by_bike
#'   - duration_by_bike
#'
#' `Kernel` indicates the kernel function from the `osmenrich` package to be
#'   used to weight the objects retrieved by their distances (or durations) from
#'   the reference objects and then convert these vectors into single numbers.
#'   The simplest kernel allows the user to count the number of occurrences
#'   of reference objects within a radius `r` and is called `kernel_uniform`.
#'
#' For more details see the introductory vignette of `osmenrich`:
#'   \code{vignette("introduction", package = "osmenrich")}
#'
#' @examples
#' \dontrun{
#' # Load libraries
#' library(tidyverse)
#' library(sf)
#'
#' # Create example dataset
#' sf_example <-
#' tribble(
#'   ~person,  ~lat,   ~lon,
#'   "Alice",  52.12,  5.09,
#'   "Bob",    52.13,  5.08,
#'   ) %>%
#'   sf::st_as_sf(
#'   coords = c("lon", "lat"),
#'   crs = 4326
#'  )
#'
#' # Enrich data creating new column `waste_baskets`
#' sf_enriched <- sf_example %>%
#'   enrich_osm(
#'     name = "n_waste_baskets",
#'     key = "amenity",
#'     value = "waste_basket",
# '    type = "points",
# '    distance = "duration_by_foot",
# '    r = 100,
# '    kernel = "uniform",
#'     reduce_fun = sum
#'   )
#' }
#'
#' @seealso [enrich_opq()]
#' @note If you want to get a large number of objects make sure to set the
#'   `.timeout` (time before request times out) and `.memsize` (maxmimum
#'   size of the request) arguments for the Overpass server and set
#'   the "max-table-size" argument correctly when starting the
#'   OSRM server(s).
#' @export
enrich_osm <- function(
                       dataset,
                       name = NULL,
                       key = NULL,
                       value = NULL,
                       type = "points",
                       measure = "spherical",
                       r = NULL,
                       kernel = "uniform",
                       reduce_fun = sum,
                       control = list(),
                       .verbose = TRUE,
                       ...) {
  if (is.null(name)) stop("Enter a query name.")
  if (length(name) > 1) {
    stop("You can enrich one query at the time only.")
  } else {
    control <- do.call("control_enrich", control)
    # Create query to OSM server
    query <- enrich_opq(
      dataset = dataset,
      name = name, key = key, value = value, type = type,
      measure = measure, r = r, kernel = kernel,
      reduce_fun = reduce_fun, control = control, .verbose = .verbose,
      ...
    )
    # Enrichment call
    enriched_data <- data_enrichment(
      ref_data = dataset, query = query, colname = name, .verbose = .verbose
    )
    return(enriched_data)
  }
}

#' @rdname enrich
#' @keywords internal
data_enrichment <- function(ref_data, query, colname, .verbose = TRUE) {
  # Check inputs
  if (!is(ref_data, "sf")) stop("Data should be sf object.")
  check_enriched_opq(query)

  # Extract the feature points and/or centroids
  # Only download points if only points are requested
  if (length(query[["type"]]) == 1 && query[["type"]] == "points") {
    attr(query, "nodes_only") <- TRUE
  }
  if (.verbose) {
    cli::cli_process_start(
      msg = cli::col_cyan(glue::glue("Downloading data for {colname}...")),
      msg_done = cli::col_green("Downloaded data for {colname}."),
      msg_failed = cli::col_red(glue::glue("Failed to download data for {colname}!"))
    )
  }

  # Retrieve data from OSM server
  ftr_data <- osmdata::osmdata_sf(q = query)
  if (.verbose) {
    cli::cli_process_done()

    cli::cli_alert_info(cli::col_cyan(sprintf(
      "Downloaded %i points, %i lines, %i polygons, %i mlines, %i mpolygons.",
      if (is.null(ftr_data$osm_points)) 0 else
        nrow(ftr_data$osm_points),
      if (is.null(ftr_data$osm_lines)) 0 else
        nrow(ftr_data$osm_lines),
      if (is.null(ftr_data$osm_polygons)) 0 else
        nrow(ftr_data$osm_polygons),
      if (is.null(ftr_data$osm_multilines)) 0 else
        nrow(ftr_data$osm_multilines),
      if (is.null(ftr_data$osm_multipolygons)) 0 else
        nrow(ftr_data$osm_multipolygons)
    )))
  }

  # Get feature sf::geometry
  first <- TRUE
  for (type in query$type) {
    geometry <- ftr_data[[paste0("osm_", type)]][["geometry"]]
    if (is.null(geometry)) next
    # Whatever the geometry, as long as not points use centroid
    # Here one could divide it depending on the geometry or choice of user
    if (type != "points") {
      geometry <- sf::st_centroid(geometry) # of_largest_polygon = T
    }
    if (first) {
      ftr_geometry <- geometry
      first <- FALSE
    } else {
      ftr_geometry <- c(ftr_geometry, geometry)
    }
  }

  if (.verbose) {
    cli::cli_process_start(
      msg = cli::col_cyan(glue::glue("Computing measure matrix for {colname}...")),
      msg_done = cli::col_green("Computed measure matrix for {colname}."),
      msg_failed = cli::col_red(glue::glue("Failed to compute measure matrix for {colname}!"))
    )
  }

  # Modify both ftr and ref to 4326
  options(warn=-1)
  ref_geometry <- sf::st_transform(ref_data, crs = 4326)
  # This command raises a warning due to different versions of GDAL
  # see: https://github.com/r-spatial/sf/issues/1419
  sf::st_crs(ftr_geometry) <- 4326
  options(warn=0)

  # Create matrix ref <-> ftr
  measure_mat <- measure_matrix(
    measure_name = query[["measure"]],
    measure_fun  = query[["measurefun"]],
    ref_geometry = ref_geometry,
    ftr_geometry = ftr_geometry
  )

  # Apply the kernel function over the rows of the measure matrix
  apply_args <-
    c(
      list(
        X = measure_mat,
        MARGIN = 1,
        FUN = query[["kernelfun"]]
      ),
      query[["kernelpars"]]
    )
  feature <- do.call(what = apply, args = apply_args)

  if (.verbose) {
    cli::cli_process_done()
    cli::cli_alert_info(cli::col_cyan(glue::glue("Adding {colname} to data.")))
  }
  ref_data[[colname]] <- feature
  return(ref_data)
}

#' @rdname enrich
#' @keywords internal
measure_matrix <- function(measure_name,
                            measure_fun,
                            ref_geometry,
                            ftr_geometry) {
  # If "spherical" then no call to OSRM necessary
  if (measure_name == "spherical") {
    matrix <- sf::st_distance(ref_geometry, ftr_geometry)
    return(matrix)
  }

  if (!check_osrm_limits(src = ref_geometry, dst = ftr_geometry)) {
    matrix <- measure_fun(ref_geometry, ftr_geometry)
  } else {
    print("Splitting main call and creating sub-calls...")
    tot_nrows <- nrow(ref_geometry) * nrow(sf::st_coordinates(ftr_geometry))
    first <- TRUE
    chunk_size <- 20000
    for (i in seq(1, tot_nrows, chunk_size)) {
      seq_size <- chunk_size
      if ((i + seq_size) > tot_nrows) seq_size <- tot_nrows - i + 1
      matrix <- measure_fun(ref_geometry[i:(i+seq_size),],
                            ftr_geometry[i:(i+seq_size),])
      if (first) {
        result <- matrix
        first <- FALSE
      } else {
        result <- rbind(result, matrix)
      }
    }
  }
}

#' @rdname enrich
#' @keywords internal
control_enrich <- function(timeout = 300, memsize = 1073741824) {
  if (!is.numeric(timeout) || timeout <= 0) {
    stop("Value of 'timeout' must be > 0")
  }
  if (!is.numeric(memsize) || memsize <= 0) {
    stop("Value of 'memsize' must be > 0")
  }
  list(timeout = timeout, memsize = memsize)
}
